File: sol_1_6_1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.ArrayList;


class sol_1_6_1 {
   private static int cycle_length(int n) {
      int length = 1;
      int tmp = n;

      while (n != 1)
      {
         if ((n % 2) == 1) // odd number
            n = n * 3 + 1;
         else           // even number
            n = n / 2;

         length++;       // keep track of cycle length
      }

      return length;
   }


   // read input from STDIN. each line must consist only of two integers.
   // the first integer is the lower bound and the second is the upper bound
   private static void input(ArrayList<Pair> list) throws FileNotFoundException {
      // read the 2 integers from file into a pair
      Scanner scanner = new Scanner(System.in);
      while (true) {
         if (scanner.hasNext()) {  // store the pair into list and keep going
            Pair pair = new Pair();
            pair._first = scanner.nextInt();
            pair._second = scanner.nextInt();
            list.add(pair);
         }
         else
            break; // exit loop if end of file is reached
      }
      scanner.close();
   }


   // process the actual entries in the array
   private static void output(ArrayList<Pair> list) {
      // for each array entry (pair of lower and upper bounds), do this:
      // for each integer from the lower bound to the upper bound, inclusive
      // of both bounds, get the cycle length for that integer, then obtain
      // the maximum value of all such cycle lengths
      for (int i = 0; i < list.size(); i++) {
         int max = 0;
         for (int j = list.get(i)._first; j <= list.get(i)._second; j++)
         {
            int tmp = cycle_length(j);
            if (tmp > max)
               max = tmp;
         }

         System.out.print(list.get(i)._first + " " + list.get(i)._second +
                      " " + max + "\n");
      }
   }


   public static void main(String[] args) throws FileNotFoundException {
      ArrayList<Pair> list = new ArrayList<Pair>();
      input(list);
      output(list);
   }
}


class Pair {
   int _first;
   int _second;
}

File: sol_1_6_1.rb
# ruby sol_1_6_1.rb in_1_6_1 | tr -d '\r' | diff out_1_6_1 -
#
# Calculate the cycle length for a given number n
def cycle_length(n)
  length = 1
  while n != 1
    # odd number
    if (n % 2) == 1
      n = n * 3 + 1
    # even number
    else
      n = n / 2
    end

    length += 1
  end

  return length;
end


def input(filename)
  pairs = []
  # read input from a file.
  File.open(filename.to_s).each do |line|
    pair = line.split(" ")
    # exit loop if end of file is reached
    if pair.size == 0
      break
    else
      # Each line must consist only of two integers. The first integer is
      # the lower bound and the second is the upper.
      # Store the pair into an array and keep going
      pairs << [pair[0].to_i, pair[1].to_i]
    end
  end

  return pairs
end


def output(pairs)
  pairs.each do |pair|
    max = 00
    # For each array entry (a pair of lower and upper bounds), do this:
    # for each integer between the lower bound and the upper bound,
    # inclusive of both bounds, get the cycle length for that integer,
    # then obtain the maximum value of all such cycle lengths
    (pair[0]..pair[1]).each do |integer|
      length = cycle_length(integer)
      if max < length
        max = length
      end
    end

    puts("#{pair[0]} #{pair[1]} #{max}")
  end
end

if ARGV.empty?
  puts("Usage: Must pass input file as command-line argument")
else
  output(input(ARGV[0]))
end
